Microsoft/mu_basecore
# 2405: Crypto Changes [Rebase & FF]

## Description

Closes #871 

All CryptoPkg changes to match the `release/202311` branch.

Updates some other packages to account for changes in libraries
used in CryptoPkg.

---

**Transition to Latest Shared Crypto**

Updates to the code to match the latest Shared Crypto code in the
release/202311 branch using the 2023.11.3 binary.

For Shared Crypto details and platform integration information,
review https://github.com/microsoft/mu_crypto_release.

Other notable changes consolidated in this commit include:

- Reconfigures OpensslLib to add elliptic curve cipher algorithms.
- Fixes a memory free bug in the HmacTest
  - Test is currently using FreePool() when it should use HmacSha256Free()
- Adds SHA384 and SHA512 to the STANDARD flavor of the crypto binary

---

Contains the following additional cherry-picked commits.

- CryptoPkg: CodeQL Fixes (all package CodeQL changes consolidated)
- [TCBZ3925] Correct the implementation of Pkcs7Sign
- Fix memory free bug in HmacTest
- Test is currently using FreePool() when it should use HmacSha256Free()
- CryptoPkg: Add host-based unit testing
- TCBZ2997: Use Non-Null CryptPkcs5Pbkdf2 for RuntimeCryptLib.inf
- Remove support for deprecated crypto from BaseCryptLib.
- TCBZ2029: Fix AARCH64 Crypto builds
- Add Variable Argument List macros for VS

---

- [x] Impacts functionality?
- [x] Impacts security?

---

Includes these 2311 commits:

- 718ec92fc526c9e30bd752bb4f4c692b7d135f60
- d6e01136697c1aed85bb83cff98b74ec11e96e1a
- 884e5da43136e0d56863ccb1c842fa74f10088ce
- 6cc02e2da5aef3364bc161854382cfada0c6a2b4
- d92ffaf8b2251e1444496dfba57f2cbb1dc43f5a
- 2ffdb64437d2d349ca8467e40846a4571a471c4b
- e9e34593db8b2e4506164ef123ccae4f5a9c699f
- 76b62ebff5cc994b56af228b961bc52b418c457c
- 0dd0d48597137cdfb1e838aaebdd46726eda6b37
- ba364a5359d283ae626fa469c397ba360e7263ed
- c6e7c62056882238ed28905bf148e271a94e005f
- 1d1fdbb611741c4abb14f53a6477c8a77f730413
- cd0963c6d5c328fddec0cda0c1561390f91cf6cb
- 7bbe2cfc0929e17cc91c0cbc93d98d077cd00a22
- 9ffc11f3e9892f045ad9c939612bd126e900a3c0
- 440050aa2823c111d956ac242c9705095c71a851
- 46a55e136cf887efd450e2e8f8ce08e1f6fb6cbc
- 9732d0d4a0a05a13e933bb9ac148ed0183f927a2
- de863fff9f3aeea689fbd9d6b639c0e5584f1f17
- ff2bb2283dd52cbcec0a86084fd530510cd9c1df
- afd1f318977d5094b2dc3ec053a2a9c480332c35
- 0437f1dd87ef3fb779be74017732909a8394aa4c
- 3bb862c0e95770aa26170a771bd57ed6bcb2c1fc
- 7cd24cd2646659729bcb0a3cfed325fe10c8b91e
- 0d9bc4121eb42d2f82bd57c91883886977de5756
- 5bbf055ab97466fc556c61e06a1bb5da6f4cc6c4
- eab839fdab7b5681ce717f4d2720d4dd03050db0
- 0f6152462f4f6d785c9f73aa7dd6ddb0fdb7420e
- abe2ebac086058e19c72be9d07abca93479abe8f
- dbddca0b5c3335d06287abfe0e6e89685c2a505d
- 6c48e1dcecabee1801921c02543126b98b04e66e
- 82b92f718921cf7164db9c18230bf3ef0464002a
- 86722c119e06c7f1ae717b37b606009db318db4c
- 2deccbe69e0705be8c7d4918d76c00f3eabda65f
- 2c4db3932afd85b73c074ca9838563e2bfc6ce36
- ec7a96c42436ed86f5c714d81add1225a4c7b165
- 984dee96cf9206df8c58ad2d05a1993bb228d62d
- 36ad94d22c5c8efd09e1b5d06b3740eb9231ad03
- 8b51b1a8136c9f48c29cb73659ab1f4f5d150776
- 326dde3a9fa4adde4d177c2078d571725bcb364a
- e71970bff938dba264db2b2d3d0427c175c79b9f
- 3a996f479eb395dfb0981dce30cf46d35f19d689
- 798cfa3b926ef2996faaba99452bd9c3624c0bda
- 674a151174b9f9f692a6f0494758f06577dcd48a
- 4474860920659f11be67b0c1c9f56b2781d6b612
- 381e74cfd03d0b4989f88708b5fabd7d638733e2
- 4748254bf0874f1c84ed2ddd94628e92fef4392b
- 4fa08d9105d7554087f1cf51e18a6bcdd60790f4
- b3a3b033319f24b80996bacd2027836c3dc073e5
- 1c141293144ea80199c4693b462af597384d0a88
- d145029f2599772f7e33d3294235eb6092cff3cd
- d1088e2ad529ab2cb5a0f7aa0e94dc8adf4fbc64
- 33e15727edb047f1f7ae715ce83259d2ea1e5548
- 3c7cc8bb7f224c735ff349f1b1186010a6e61aa4
- eafd7ac5d1b60861306165b64212198e9a14fb39
- 5c597f57bfc59b73a95730af3d4782ddf91eac4a
- 6d509b255ed3bcb9457bf0953ff558a384f12846
- 25a5997dbdfdc23eb2dbf89dda5d8918d842b3ea
- ca1455a6a04e819ec8c2057495b720b9550144ec

## How This Was Tested

- Changes from `release/202311`

## Integration Instructions

- See the shared crypto repo for more details about shared crypto
  and integration information https://github.com/microsoft/mu_crypto_release
